<?xml version="1.0"?>
<robot xmlns:xacro="http://wiki.ros.org/xacro" name="ur3e" >

  <material name="white02">
    <color rgba="1 1 1 0.2"/>
  </material>

  <!-- parameters -->
  <xacro:arg name="gazebo" default="false"
             doc="whether robot is loaded in gazebo simulator"/>
  <xacro:arg name="joint_limit_params"
             default="$(find ur_description)/config/ur3e/joint_limits.yaml"/>
  <xacro:arg name="kinematics_params"
             default="$(find agimus_demos)/ur3/pointing/config/calibrated-params.yaml"/>
  <xacro:arg name="physical_params"
             default="$(find ur_description)/config/ur3e/physical_parameters.yaml"/>
  <xacro:arg name="visual_params"
             default="$(find ur_description)/config/ur3e/visual_parameters.yaml"/>
  <!-- legal values:
       - hardware_interface/PositionJointInterface
       - hardware_interface/VelocityJointInterface
       - hardware_interface/EffortJointInterface
-->
  <xacro:arg name="transmission_hw_interface"
             default="hardware_interface/PositionJointInterface"/>
  <xacro:arg name="safety_limits" default="false"/>
  <xacro:arg name="safety_pos_margin" default="0.15"/>
  <xacro:arg name="safety_k_position" default="20"/>

  <!-- Offset of the tip link (obtained by calibration) -->
  <xacro:arg name="tip_offset_x" default="0.0000"/>
  <xacro:arg name="tip_offset_y" default="0.0000"/>
  <xacro:arg name="tip_offset_z" default="0.0000"/>

  <!-- import gazebo plugins macro -->
  <xacro:if value="$(arg gazebo)">
    <xacro:include filename="$(find agimus_demos)/urdf/d435_gazebo.urdf.xacro" />
  </xacro:if>
  <!-- import main macro -->
  <xacro:include filename="$(find ur_description)/urdf/inc/ur_macro.xacro"/>
  
  <!-- arm -->
  <xacro:ur_robot
      prefix=""
      joint_limits_parameters_file="$(arg joint_limit_params)"
      kinematics_parameters_file="$(arg kinematics_params)"
      physical_parameters_file="$(arg physical_params)"
      visual_parameters_file="$(arg visual_params)"
      transmission_hw_interface="$(arg transmission_hw_interface)"
      safety_limits="$(arg safety_limits)"
      safety_pos_margin="$(arg safety_pos_margin)"
      safety_k_position="$(arg safety_k_position)"/>

  <!-- load parameters of camera pose -->
  <xacro:property name="camera_pose_file"
                  value="$(find agimus_demos)/ur3/pointing/config/calibrated-params.yaml" />
  <xacro:property name="camera_pose" value="${load_yaml(camera_pose_file)}"/>
  
   <link name="world" />

    <link name="support_link">
      <inertial>
        <mass value="200"/>
	<origin xyz="0 0 0.44"/>
        <inertia ixx="50"  ixy="0"  ixz="0" iyy="50" iyz="0" izz="50" />
      </inertial>
      
      <!-- table -->
      <visual>
	<origin xyz="0.0 -0.305 0.5"/>
        <geometry>
	        <box size="0.80 1 1"/>
        </geometry>
      </visual>
      <collision>
        <origin xyz="0.0 -0.305 0.5"/>
        <geometry>
	        <box size="0.80 1 1"/>
        </geometry>
      </collision>
      
      <!-- support -->
     <visual>
       <origin xyz="0 0 1.01"/>
       <geometry>
	        <box size="0.16 0.16 0.02"/>
       </geometry>
     </visual>
     <collision>
       <origin xyz="0 0 1.01"/>
       <geometry>
	        <box size="0.16 0.16 0.02"/>
       </geometry>
     </collision>     
    </link>

     <link name="walls">
     <!-- front wall -->
     <!-- visual -->
	<!-- origin xyz="0.385 -0.305 0.37"/ -->
        <!-- geometry -->
	  <!-- box size="0.03 1 0.74"/ -->
        <!-- /geometry -->
	<!-- material name="white02"/ -->
     <!-- /visual -->
      <collision>
       <origin xyz="0.385 -0.305 0.37"/>
       <geometry>
	 <box size="0.03 1 0.74"/>
       </geometry>
      </collision>
      
      <!-- behind wall -->
     <visual>
	<origin xyz="-0.385 -0.305 0.37"/>
        <geometry>
	  <box size="0.03 1 0.74"/>
        </geometry>
	<material name="white02"/>
     </visual>
      <collision>
       <origin xyz="-0.385 -0.305 0.37"/>
       <geometry>
	 <box size="0.03 1 0.74"/>
       </geometry>
      </collision>

      <!-- left wall -->
     <visual>
	<origin xyz="0 -0.79 0.37"/>
        <geometry>
	  <box size="0.74 0.03 0.74"/>
        </geometry>
	<material name="white02"/>
     </visual>
      <collision>
       <origin xyz="0 -0.79 0.37"/>
       <geometry>
	 <box size="0.74 0.03 0.74"/>
       </geometry>
      </collision>

     <!-- right wall -->
     <visual>
	<origin xyz="0 0.185 0.37"/>
        <geometry>
	  <box size="0.74 0.03 0.74"/>
        </geometry>
	<material name="white02"/>
     </visual>
      <collision>
       <origin xyz="0 0.185 0.37"/>
       <geometry>
	 <box size="0.74 0.03 0.74"/>
       </geometry>
      </collision>

      <!-- above wall -->
     <visual>
	<origin xyz="0 -0.305 0.755"/>
        <geometry>
	  <box size="0.80 1 0.03"/>
        </geometry>
	<material name="white02"/>
     </visual>
      <collision>
       <origin xyz="0 -0.305 0.75"/>
       <geometry>
	 <box size="0.80 1 0.03"/>
       </geometry>
     </collision>
    </link> 


<joint name="walls_joint" type="fixed">
     <parent link="support_link" />
     <child link = "walls" />
      <origin xyz="0.0 0.0 1" rpy="0.0 0.0 0" />
   </joint>

   <joint name="support_joint" type="fixed">
     <parent link="world" />
     <child link = "support_link" />
     <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
   </joint>
  
   
   <joint name="robot_joint" type="fixed">
     <parent link="support_link" />
     <child link = "base_link" />
      <origin xyz="0.0 0.0 1" rpy="0.0 0.0 -1.5707963" />
   </joint>
   
    <!-- End Effector -->
    <joint name="ee_fixed_joint" type="fixed">
        <parent link="wrist_3_link"/>
        <child link="ee_link"/>
        <origin rpy="0.0 0.0 1.57079632679" xyz="0.0 0.0823 0.0"/>
    </joint>
    <link name="ee_link">
    <collision>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <origin rpy="0 0 0" xyz="-0.01 0 0"/>
    </collision>
    </link>
    
        
  <!-- Robotiq Coupler -->
  <!-- Gripper -->
  <xacro:include filename="$(find agimus_demos)/ur3/pointing/urdf/robotiq_gripper/robotiq_85_gripper.urdf.xacro" />
  
  <!--  + Height added by the coupler: 8mm -->
  <!--  + Reference frame: at the middle (4mm) -->
  <link name="robotiq_coupler">
    <visual>
      <geometry>
        <mesh filename="package://agimus_demos/ur3/pointing/meshes/robotiq_85_coupler.stl" />
      </geometry>
      <material name="flat_black"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://agimus_demos/ur3/pointing/meshes/robotiq_85_coupler.stl" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="2.073e-05   1.45286e-03  -1.1049e-03" rpy="0 0 0" />
      <mass value="0.168" />
      <inertia ixx="6.69695624e-05" ixy="5.21511788e-09" ixz="-2.74383009e-08" 
               iyy="7.85088161e-05" iyz="5.41105193e-07" izz="1.41819717e-04"/>
    </inertial>
  </link>
  
  <joint name="robotiq_coupler_joint" type="fixed">
    <origin xyz="0 0 0.004" rpy="0 0 ${-pi/2.0}" />
    <parent link="tool0"/>
    <child link="robotiq_coupler"/>
  </joint>
  
  <gazebo reference="robotiq_coupler">
    <mu1>0.9</mu1>
    <mu2>0.9</mu2>
    <material>Gazebo/FlatBlack</material>
  </gazebo>

  <!-- Attach the robotiq 85 gripper -->
  <xacro:robotiq_85_gripper prefix="" parent="robotiq_coupler" >
    <origin xyz="0 0 0.004" rpy="0 ${-pi/2} ${pi}"/>
  </xacro:robotiq_85_gripper> 
   
   <!-- Camera -->
   <link name="ref_camera_link">
   </link>

    <joint name="ref_camera_joint" type="fixed">
     <parent link="wrist_3_link" />
     <child link = "ref_camera_link" />
     <origin xyz="${camera_pose['camera']['x']}
                  ${camera_pose['camera']['y']}
                  ${camera_pose['camera']['z']}"
	     rpy="${camera_pose['camera']['roll']}
                  ${camera_pose['camera']['pitch']}
                  ${camera_pose['camera']['yaw']}"/>
   </joint>

   <xacro:include filename="$(find realsense2_description)/urdf/_d435.urdf.xacro" />

   <xacro:sensor_d435 parent="ref_camera_link" name="camera" use_nominal_extrinsics="true">
     <!-- <origin xyz="0 0.115 0" rpy="0 -1.0472 -1.57079632679"/> -->
      <origin xyz="0 0 0" rpy="0 0 0"/>
   </xacro:sensor_d435>
</robot>
